File: 2021/10/02/main.go
package main

func calculateMinimumHP(dungeon [][]int) int {
	n, m := len(dungeon[0]), len(dungeon)
	if m == 0 || n == 0 {
		return 1
	}

	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
		for j := range dp[i] {
			dp[i][j] = 1<<63 - 1
		}
	}

	dp[m][n-1] = 1
	for i := m - 1; i >= 0; i-- {
		for j := n - 1; j >= 0; j-- {
			mn := min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j]
			dp[i][j] = max(mn, 1)
		}
	}

	return dp[0][0]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 2021/10/02/main.py
class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        n, m = len(dungeon[0]), len(dungeon)
        if m == 0 or n == 0:
            return 1
        dp = [[1<<63 - 1] * (n+1) for _ in range(n+1)]
        dp[m][n-1] = 1
        for i in range(m-1, -1, -1):
            for j in range(n-1, -1, -1):
                mn = min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j]
                dp[i][j] = max(mn, 1)

        return dp[0][0]